### PR TITLE
update garden cluster kubeconfig in gardenlet on CA rotation

### PR DESCRIPTION
**Pull request by Diaphteiros**
**How to categorize this PR?**

/area security
/kind enhancement

**What this PR does / why we need it**:
Before this change, the gardenlet would only create a new kubeconfig for accessing the garden cluster if it couldn't find one that had been created by a previous bootstrapping process. This way, gardenlet would not notice if the CA of the kube-apiserver changed.

Now, the bootstrapping process - which generates a new kubeconfig for accessing the garden cluster - is also triggered if an existing kubeconfig is found but its CA differs from the CA in the provided bootstrapping kubeconfig.

**Release note**:

```other operator
Gardenlet will now update its kubeconfig if `gardenClientConnection.gardenClusterCACert` is specified and contains a different CA cert than the one currently used in the kubeconfig.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
I'm confused, this PR just contains a log message change?
/assign @Diaphteiros

**Comment by Diaphteiros** [comment]
@rfranzke Oh, somehow some changes got lost o.O

**Comment by Diaphteiros** [comment]
OK, looks better now. I copied the project from a temporary folder into the GOPATH, but forgot to update my editor, so I modified the version in the temporary folder but committed the one from the GOPATH -.-

File: cmd/gardenlet/app/bootstrap.go [hunk 1]
```diff
@@ -44,15 +47,18 @@ func bootstrapKubeconfig(
 		return nil, "", "", err
 	}
 
-	if len(gardenKubeconfig) > 0 {
-		logger.Info("Found kubeconfig generated from bootstrap process. Using it")
-		return gardenKubeconfig, "", "", nil
+	gardenKubeconfigFound := len(gardenKubeconfig) > 0
+	if gardenKubeconfigFound {
+		logger.Info("Found kubeconfig generated from bootstrap process. Fetching boostrap kubeconfig to check for changed kube-apiserver CA.")
```
Comment (rfranzke) [inline cmd/gardenlet/app/bootstrap.go]:
```suggestion
		logger.Info("Found kubeconfig generated from bootstrap process. Fetching bootstrap kubeconfig to check for changed kube-apiserver CA.")
```

File: cmd/gardenlet/app/bootstrap.go [hunk 2]
```diff
@@ -44,15 +47,18 @@ func bootstrapKubeconfig(
...
+	} else {
+		logger.Info("No kubeconfig from a previous bootstrap found. Starting bootstrap process.")
```
Comment (rfranzke) [inline cmd/gardenlet/app/bootstrap.go]:
```suggestion
		logger.Info("No kubeconfig from a previous bootstrap procedure found. Starting bootstrap process.")
```

File: cmd/gardenlet/app/bootstrap.go [hunk 3]
```diff
@@ -44,15 +47,18 @@ func bootstrapKubeconfig(
...
 	}
 
-	logger.Info("No kubeconfig from a previous bootstrap found. Starting bootstrap process.")
-
 	if config.GardenClientConnection.BootstrapKubeconfig == nil {
-		logger.Warn("Unable to perform kubeconfig bootstrapping. The gardenlet configuration `.gardenClientConnection.bootstrapKubeconfig` is not set")
+		logger.Warn("Unable to fetch boostrap kubeconfig. The gardenlet configuration `.gardenClientConnection.bootstrapKubeconfig` is not set")
```
Comment (rfranzke) [inline cmd/gardenlet/app/bootstrap.go]:
```suggestion
		logger.Warn("Unable to fetch bootstrap kubeconfig. The gardenlet configuration `.gardenClientConnection.bootstrapKubeconfig` is not set")
```

Comment (rfranzke) [reply cmd/gardenlet/app/bootstrap.go]: This log should probably only be written if the garden kubeconfig was not found?

**Review by rfranzke — COMMENTED** [review]
I find this code and the new `if gardenKubeconfigFound { return }` statements somewhat confusing. Instead of modifying this entire bootstrapping code, can't you just pass the CA cert to https://github.com/gardener/gardener/blob/e2708c5cd905f8ff79fdc599989321f6ca873fac/charts/gardener/gardenlet/values.yaml#L36 and exchange it in the existing kubeconfig if it is different? This way, no new bootstrap procedure is needed and the code probably gets a bit cleaner?

**Comment by Diaphteiros** [comment]
@rfranzke That's a nice idea, I didn't think of that. Will try to implement it that way.

File: pkg/gardenlet/bootstrap/util/util.go [hunk 4]
```diff
@@ -105,6 +106,42 @@ func UpdateGardenKubeconfigSecret(ctx context.Context, certClientConfig *rest.Co
 	return kubeconfig, nil
 }
 
+// UpdateGardenKubeconfigCAIfChanged checks if the garden cluster CA given in the gardenClientConnection differs from the CA in the kubeconfig secret
+// and updates the secret to contain the new CA if that's the case.
+func UpdateGardenKubeconfigCAIfChanged(ctx context.Context, seedClient client.Client, kubeconfigAsBytes []byte, gardenClientConnection *config.GardenClientConnection) ([]byte, error) {
+	if kubeconfigAsBytes == nil {
+		return nil, fmt.Errorf("no kubeconfig given")
+	}
```
Comment (rfranzke) [inline pkg/gardenlet/bootstrap/util/util.go:114]:
```suggestion
	}

```

(please introduce similar newlines to improve readability)

File: pkg/gardenlet/bootstrap/util/util.go [hunk 5]
```diff
@@ -105,6 +106,42 @@ func UpdateGardenKubeconfigSecret(ctx context.Context, certClientConfig *rest.Co
 	return kubeconfig, nil
 }
 
+// UpdateGardenKubeconfigCAIfChanged checks if the garden cluster CA given in the gardenClientConnection differs from the CA in the kubeconfig secret
+// and updates the secret to contain the new CA if that's the case.
+func UpdateGardenKubeconfigCAIfChanged(ctx context.Context, seedClient client.Client, kubeconfigAsBytes []byte, gardenClientConnection *config.GardenClientConnection) ([]byte, error) {
```
Comment (rfranzke) [inline pkg/gardenlet/bootstrap/util/util.go]:
```suggestion
func UpdateGardenKubeconfigCAIfChanged(ctx context.Context, seedClient client.Client, kubeconfig []byte, gardenClientConnection *config.GardenClientConnection) ([]byte, error) {
```

Comment (rfranzke) [inline pkg/gardenlet/bootstrap/util/util.go]: Please add a unit test for this function

**Review by rfranzke — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Comment by acumino** [comment]
@rfranzke should the release not in this be breaking and not others?

**Comment by rfranzke** [comment]
No, why?

**Comment by acumino** [comment]
> No, why?

Not really sure but ca is changed in kubeconfig and if someone is using kubeconfig ca for some other use?

**Comment by rfranzke** [comment]
Nobody except gardenlet itself should use gardenlet's kubeconfig

**Pull request merged**